### ISSUE TITLE
Object rest and spread properties are not supported by Electron with Chrome 59

### ISSUE THREAD

**Issue by godban** [cypress-io/cypress#2069]
Hey guys, could you update Chrome in your Electron, please? Next version has the normal support of Object rest and spread properties. Now we have to create a separate build for Cypress.

### Current behavior:

Tests failed because of unknown syntax

### Desired behavior:

Tests work, syntax recognized

### Steps to reproduce:

1. Use object rest and spread properties in code
2. Run tests in headless mode

### Versions

Cypress 3.0.2

**Comment by brian-mann** [comment]
Sure, we can do that. Curious why you wouldn't just transpile your test code using babel/webpack?

**Comment by godban** [comment]
We actually doing that now. But this takes build time, extra configuration, extra jobs and extra dependencies that could be omitted.

**Comment by edouard-lopez** [comment]
Still getting the issue with `"cypress": "^3.1.2"`.

Is there a `babel` configuration to circumvent this ?

**Comment by edouard-lopez** [comment]
Still having the issue despite installing [`@babel/plugin-transform-spread`](https://babeljs.io/docs/en/babel-plugin-transform-spread)

    yarn add --dev @babel/plugin-transform-spread browserslist

**package.json/scripts**

    "@babel/plugin-transform-spread": "^7.0.0",
    "browserslist": "^4.3.4",

**package.json/babel**
```
  "babel": {
    "presets": [
      "@babel/preset-env"
    ],
    "plugins": [
      "@babel/plugin-transform-spread"
    ]
  },
```

**package.json/browserslist**
```
  "browserslist": [
    "Electron >= 1.8",
    ">0.2%",
    "not dead",
    "not ie <= 11",
    "not op_mini all"
  ]
```

**Comment by edouard-lopez** [comment]
### Solution with `--browser`

I switch to my _Linux Mint_ laptop and was able to run the test with the `--browser chromium` option. However I want to run my tests on travis without installing an [extra browser](https://docs.travis-ci.com/user/reference/trusty/#headless-browser-testing-tools) and Chromium isn't available by default. 

### Solution with [`uncaught:exception` app event](uncaught:exception)
So I went for the solution to disable uncaught exception from app in `cypress/support/index.js`:

```
Cypress.on("uncaught:exception", (err, runnable) => {
  console.debug(">> uncaught:exception disabled in cypress/support/index.js");
  return false;  // prevents Cypress from failing the test
});
```

My **app doesn't throw any exception**, either in development or production mode, when in run in the browser. Only `cypress` complain

@jennifer-shehane, @brian-mann  I'm not really fund of this solution as it disable all exceptions. Could you point to a working `babel@7`+`webpack@4` configuration?

**Comment by jennifer-shehane** [comment]
There are examples of webpack preprocessor configs in our https://github.com/cypress-io/cypress-webpack-preprocessor repo

**Comment by cake-jnorman** [comment]
@jennifer-shehane 

I've seen you post that link across multiple issues referring to the spread operator and cypress-webpack.  I am also having this issue.  

The Cypress server won't even start.  My code is being transpiled and it runs just fine in both dev and production builds.  but when i try to run cypress coupled with the cypress-webpack-preprocessor i get the "SyntaxError: Unexpected token ..." error (the spread operator), referring to a line in my webpack config file.  The fact that several people are asking about this particular issue probably means that the documentation alone is enough.  

Is there any chance that we can get get a bit of elaboration on this issue?  Cypress is awesome.  This issue however, is driving me mad.  (Yes, i may just be an idiot... but it appears that i am not alone)

Thank you!

**Comment by donald-s** [comment]
@cake-jnorman If I'm understanding your problem correctly, it sounds to me like one of your Cypress tests is requiring a file that contains a spread operator. As mentioned above, Electron (or maybe just this version of Electron - I'm not entirely sure) doesn't support the spread operator. So, when Cypress tries to require the file, it throws that error.

Other than the workaround you mentioned, you can solve this issue by using one of the Docker images that contains Chrome (cypress/browsers:chrome67 in my case) and telling Cypress to use Chrome with the `--browser chrome` option: `yarn cypress run --browser chrome`. Unfortunately, this option doesn't support screen recordings, so that's something to keep in mind.

**Comment by nachocab** [comment]
> Sure, we can do that. Curious why you wouldn't just transpile your test code using babel/webpack?

@brian-mann I'm running into this issue with Typescript after using [your recipe](https://github.com/cypress-io/cypress-example-recipes/blob/master/examples/preprocessors__typescript-webpack/webpack.config.js). It says that webpack-preprocessor handless babel, so I don't understand why I get the Error: `Unexpected token ...`

**Comment by imranismail** [comment]
I'm still facing this problem.

I didn't do anything fancy but just added a new command using existing commands from cypress, and this error pops up.

The following code snippet is the custom command that I've added on top of a fresh project:

```js
Cypress.Commands.add('login', (code, phone) => {
  cy.visit('/');
  cy.getCookie('accessToken').should('be.null');
  cy.contains('a', 'account_circle').click();
  cy.url().should('include', 'login');
  cy.contains('a', 'Continue with phone number').click();
  cy.url().should('include', 'auth');
  cy.get('select[name="countryCode"]').select(code);
  cy.get('input[name="phone"]').type(phone));
  cy.contains('button', 'Continue').click();
  cy.url().should('include', 'verify');
  cy.get('input[name=code]').type('1337');
  cy.get('button').contains('Verify').click();
  cy.url().should('not.include', 'login')
          .and('not.include', 'verify')
          .and('not.include', 'auth')

  return cy.getCookie('accessToken');
})
```

**Comment by kabukki** [comment]
In my case, I transpiled everything correctly, the "..." syntax came from a vendor bundle.

**Comment by godban** [comment]
How hard Cypress coupled to this version of Electron?

Because web always moves on and sticking with this version will bring more and more issues in future.

**Comment by jennifer-shehane** [comment]
Electron upgrade is planned to be released in [version 4.0](https://github.com/cypress-io/cypress/issues/2840)

**Comment by charlex** [comment]
@jennifer-shehane Are there any updates on the release of 4.0? (I don't see 4.0 mentioned on the [roadmap](https://docs.cypress.io/guides/references/roadmap.html).)

**Comment by tuxd** [comment]
`ts-loader` module itself which runs inside the cypress electron js context requires the `{..., nonsense}`

I want my lief back!!

**Comment by pikadanx** [comment]
@tuxd I ran into the same issue with `ts-loader@6.0.0`.

Downgrading to `ts-loader@5.4.5` solved the spread operator issue for me.

**Comment by PinkyJie** [comment]
me too, I upgraded ts-loader to version 6, then got the same issue.

**Comment by brian-mann** [comment]
We recently upgraded Electron in the last release (today) and spent a lot of time upgrading the internal network layer to accommodate the next series of Electron upgrades.

**Comment by godban** [comment]
@brian-mann great news! thank you

**Comment by Manny91** [comment]
How can I update the version that Cypress latest release(3.2.0) is using so I can fix it?

**Comment by Asjas** [comment]
@Manny91 The latest Cypress release is `3.3.1` and is based on Chromium 74. So just update the npm package to the latest release.

**Comment by flotwig** [comment]
@Asjas Cypress 3.3.1 bundles Electron 2, which is the equivalent of Chromium 64.

**Comment by sainthkh** [comment]
@jennifer-shehane 

It is now solved in Cypress 3.8.3. I guess it might have been solved from 3.8.0 (Electron 7.x update).

Test:

```js
it('test', () => {
  const aa = {
    a: 'str',
    b: 1234,
  }

  const bb = {
    ...aa,
    c: 333,
  }

  const { a, ...z } = bb

  expect(bb.b).to.eq(1234)
  expect(a).to.eq('str')
  expect(z).to.deep.eq({
    b: 1234,
    c: 333,
  })
})
```

**Comment by jennifer-shehane** [comment]
Closing as resolved. 

If you're experiencing a bug similar to this in Cypress, please [open a new issue](https://github.com/cypress-io/cypress/issues/new) with a fully reproducible example that we can run. There may be a specific edge case with the issue that we need more detail to fix.
